l5kit/l5kit/simulation/unroll.py

```diff
@@ -0,0 +1,226 @@
+        for a given scene dataset
+
+        :param scene_id: the scene indices
+        :param sim_dataset: the simulation dataset
+        """
+        if scene_id not in sim_dataset.scene_indices:
+            raise ValueError(f"scene: {scene_id} not in {sim_dataset.scene_indices}")
+
+        self.scene_id = scene_id
+        self.recorded_dataset = sim_dataset.recorded_scene_dataset_batch[scene_id]
+        self.simulated_dataset = sim_dataset.scene_dataset_batch[scene_id]
+
+        self.recorded_ego_states = self.recorded_dataset.dataset.frames
+        self.recorded_agents_states = self.recorded_dataset.dataset.agents
+        self.simulated_ego_states = self.simulated_dataset.dataset.frames
+        self.simulated_agents_states = self.simulated_dataset.dataset.agents
+
+    def get_scene_id(self) -> int:
+        """
+        Get the scene index for this SimulationOutputs
+
+        :return: the scene index
+        """
+        return self.scene_id
+
+
+class SimulationLoop:
+    def __init__(self, sim_cfg: SimulationConfig, dataset: EgoDataset,
+                 model_ego: Optional[torch.nn.Module] = None,
+                 model_agents: Optional[torch.nn.Module] = None):
+        """
+        Create a simulation loop object capable of unrolling ego and agents
+        :param sim_cfg: configuration for unroll
+        :param dataset: EgoDataset used while unrolling
+        :param model_ego: the model to be used for ego
+        :param model_agents: the model to be used for agents
+        """
+        self.sim_cfg = sim_cfg
+        if not sim_cfg.use_ego_gt and model_ego is None:
+            raise ValueError("ego model should not be None when simulating ego")
+        if not sim_cfg.use_agents_gt and model_agents is None:
+            raise ValueError("agents model should not be None when simulating agent")
+
+        self.model_ego = model_ego
+        self.model_agents = model_agents
+        self.dataset = dataset
+
+    def unroll(self, scene_indices: List[int]) -> List[SimulationOutputs]:
+        """
+        Simulate the dataset for the given scene indices
+        :param scene_indices: the scene indices we want to simulate
+        :return: the simulated dataset
+        TODO: this should probably return something else
+        """
+        sim_dataset = SimulationDataset(self.dataset, scene_indices, self.sim_cfg.start_frame_index,
+                                        self.sim_cfg.disable_new_agents, self.sim_cfg.distance_th_far,
+                                        self.sim_cfg.distance_th_close)
+
+        if self.sim_cfg.num_simulation_steps is None:
+            range_unroll = range(self.sim_cfg.start_frame_index, len(sim_dataset))
+        else:
+            end_frame_index = min(len(sim_dataset), self.sim_cfg.num_simulation_steps + self.sim_cfg.start_frame_index)
+            range_unroll = range(self.sim_cfg.start_frame_index, end_frame_index)
+
+        for frame_index in tqdm(range_unroll):
```

Comment:
It would be nice if you can add a flag on the config like `show_progress` to enable/disable `tqdm` here.

Reply:
which config? the simulation one?